test/javascript/turbolinks_replace_test.coffee

```diff
@@ -0,0 +1,158 @@
+assert = chai.assert
+
+suite 'Turbolinks.replace()', ->
+  setup (done) ->
+    @iframe = document.createElement('iframe')
+    @iframe.style.display = 'none'
+    @iframe.setAttribute('src', 'iframe.html')
+    document.body.appendChild(@iframe)
+    @iframe.onload = =>
+      @window = @iframe.contentWindow
+      @document = @window.document
+      @Turbolinks = @window.Turbolinks
+      @$ = (selector) => @document.querySelector(selector)
+      done()
+
+  teardown ->
+    document.body.removeChild(@iframe)
+
+  test "default", (done) ->
+    doc = """
+      <!DOCTYPE html>
+      <html>
+      <head>
+        <meta charset="utf-8">
+        <title>new title</title>
+        <meta content="new-token" name="csrf-token">
+        <script>var headScript = true</script>
+      </head>
+      <body new-attribute>
+        <div id="new-div"></div>
+        <div id="permanent" data-turbolinks-permanent>new content</div>
+        <div id="temporary" data-turbolinks-temporary>new content</div>
+        <script>var bodyScript = true</script>
+        <script data-turbolinks-eval="false">var bodyScriptEvalFalse = true</script>
+      </body>
+      </html>
+    """
+    body = @$('body')
+    permanent = @$('#permanent')
+    beforeUnloadFired = false
+    @document.addEventListener 'page:before-unload', =>
+      assert.notOk @window.bodyScript
+      assert.notOk @$('#new-div')
+      assert.notOk @$('body').hasAttribute('new-attribute')
+      assert.ok @$('#div')
+      assert.equal @$('meta[name="csrf-token"]').getAttribute('content'), 'token'
+      assert.equal @document.title, 'title'
+      assert.equal @$('body'), body
+      beforeUnloadFired = true
+    @document.addEventListener 'page:change', =>
+      assert.ok beforeUnloadFired
+      assert.ok @window.bodyScript
+      assert.notOk @window.headScript
+      assert.notOk @window.bodyScriptEvalFalse
+      assert.ok @$('#new-div')
+      assert.ok @$('body').hasAttribute('new-attribute')
+      assert.notOk @$('#div')
+      assert.equal @$('#permanent').textContent, 'permanent content'
+      assert.equal @document.title, 'new title'
+      assert.equal @$('meta[name="csrf-token"]').getAttribute('content'), 'new-token'
+      assert.notEqual @$('#permanent'), permanent # permanent nodes are cloned
```

Comment:
Yeah, would be great if we weren't exposing this to potential memory leaks. AdoptNode sounds like what we want.